### PR TITLE
Remove deprecated function bindToContext

### PR DESCRIPTION
**Pull request by markfields**
# See [AB#385](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/385)

## Description

`bindToContext` is deprecated on both `IFluidDataStoreChannel` and `IFluidDataStoreContext`.  Their replacements were added a while ago as optional, so make those required now and remove fakllback code that called `bindToContext`.

## Steps to Reproduce Bug and Validate Solution

If CI passes we are good.  Partners were alerted that this will be removed in the major bump following 1.0.

## PR Checklist

> Use the check-list below to ensure your branch is ready for PR. If the item is not applicable, leave it blank.

-   [x] I have updated the documentation accordingly.
-   [x] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] My code follows the code style of this project.
-   [x] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [x] Yes
-   [ ] No

This has been planned for many releases, primary partners are aware of removal timeline, which was also communicated in 1.0's BREAKING.md.

## Testing

No change in current behavior.

@agarwal-navin - Are there compat tests for ContainerRuntime/DataStoreRuntime?  If so they should only do n/n-1 I suppose so we should be good.

### CODE REVIEW

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 1]
```diff
@@ -281,14 +278,6 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         this._attachState = this.containerRuntime.attachState !== AttachState.Detached && this.existing ?
             this.containerRuntime.attachState : AttachState.Detached;
 
-        this.bindToContext = () => {
-            assert(this.bindState === BindState.NotBound, 0x13b /* "datastore context is already in bound state" */);
```
Comment (scarlettjlee) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:285]: Not really related to this PR specifically, but does something clean up these strings once they are no longer used?

Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:285]: Good questions. @alexvy86?

Comment (alexvy86) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:285]: Not right now. A recent change of mine started keeping track of the mapping from code to original message in a `.ts` file in `@fluidframework/test-runtime-utils` but I didn't change what `policy-check` already did in the line itself (just move the string to a comment). Since we now have the mapping file I guess the comments are technically unnecessary, but I'd find them convenient for when others read the code.

Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:285]: Yeah definitely agree the comments are convenient and should stick around

**Comment by markfields** [comment]
(still need to update tests - they don't even build yet)

File: BREAKING.md [hunk 2]
```diff
@@ -14,6 +14,18 @@ There are a few steps you can take to write a good change note and avoid needing
...
+# 2.0.0
+
+## 2.0.0 Upcoming changes
+
+## 2.0.0 Breaking changes
+- [bindToContext removed from IFluidDataStoreChannel and IFluidDataStoreContext](#bindToContext-removed-from-IFluidDataStoreChannel-and IFluidDataStoreContext)
+
+### bindToContext removed from IFluidDataStoreChannel and IFluidDataStoreContext
+`bindToContext` has been removed from `IFluidDataStoreChannel` and `IFluidDataStoreChannel` in the next major release.
+Please use `IFluidDataStoreChannel.makeVisibleAndAttachGraph()` and `IFluidDataStoreChannel.makeLocallyVisible()` instead.
```
Comment (agarwal-navin) [inline BREAKING.md:26]: There are internal APIs and should not be used. There is no alternative for this - just don't use it :)

File: BREAKING.md [hunk 3]
```diff
@@ -14,6 +14,18 @@ There are a few steps you can take to write a good change note and avoid needing
...
 
+# 2.0.0
+
+## 2.0.0 Upcoming changes
+
+## 2.0.0 Breaking changes
+- [bindToContext removed from IFluidDataStoreChannel and IFluidDataStoreContext](#bindToContext-removed-from-IFluidDataStoreChannel-and IFluidDataStoreContext)
+
+### bindToContext removed from IFluidDataStoreChannel and IFluidDataStoreContext
+`bindToContext` has been removed from `IFluidDataStoreChannel` and `IFluidDataStoreChannel` in the next major release.
```
Comment (agarwal-navin) [inline BREAKING.md:25]: Remove "in the next major release".

Comment (agarwal-navin) [reply BREAKING.md:25]:
```suggestion
`bindToContext` has been removed from `IFluidDataStoreChannel` and `IFluidDataStoreContext`.
```

File: api-report/container-runtime-definitions.api.md [hunk 4]
```diff
@@ -68,21 +67,12 @@ export interface IContainerRuntimeEvents extends IContainerRuntimeBaseEvents {
     (event: "localHelp", listener: (message: IHelpMessage) => void): any;
 }
 
-// @public @deprecated (undocumented)
-export interface IDataStoreWithBindToContext_Deprecated extends IDataStore {
```
Comment (agarwal-navin) [inline api-report/container-runtime-definitions.api.md:72]: Do we need to add this to breaking.md too? FFX is using this interface.

File: packages/framework/aqueduct/package.json [hunk 5]
```diff
@@ -103,6 +103,11 @@
   "module:es5": "es5/index.js",
   "typeValidation": {
     "version": "1.0.0",
-    "broken": {}
+    "broken": {
```
Comment (agarwal-navin) [inline packages/framework/aqueduct/package.json:106]: Looks unrelated to this change.

Comment (markfields) [reply packages/framework/aqueduct/package.json:106]: No it's from these type changes trickling up to this type. You can look at the old pipeline run for the error message if you're curious.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 6]
```diff
@@ -446,12 +446,12 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
      * 1. Sending an Attach op that includes all existing state
      * 2. Attaching the graph if the data store becomes attached.
      */
-    public bindToContext() {
```
Comment (agarwal-navin) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:449]: `bindToContext` should be removed and the contents should be moved to `makeVisibleAndAttachGraph`. This will ensure that existing calls to `bindToContext` by casting to any would not work. `bindState` will also be no longer needed.

**Comment by agarwal-navin** [comment]
> @agarwal-navin - Are there compat tests for ContainerRuntime/DataStoreRuntime? If so they should only do n/n-1 I suppose so we should be good.

The specific tests for this are here - https://github.com/microsoft/FluidFramework/blob/main/packages/test/test-end-to-end-tests/src/test/newFluidObjectVisibility.spec.ts. They should be updated to do full compat now.

**Comment by agarwal-navin** [comment]
Another test that should be moved to be full compat - https://github.com/microsoft/FluidFramework/blob/68c2bc5a2b4ee3ae9acb8f42c0ed966d366c4d19/packages/test/test-end-to-end-tests/src/test/detachedContainerTests.spec.ts#L693

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 6 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:449]: +1

Comment (markfields) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:449]: I went ahead and removed BindState enum too then.

**Review by vladsud — APPROVED** [review]
I only briefly glanced at it

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 7]
```diff
@@ -281,14 +278,6 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         this._attachState = this.containerRuntime.attachState !== AttachState.Detached && this.existing ?
             this.containerRuntime.attachState : AttachState.Detached;
 
-        this.bindToContext = () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:284]: Restore this, for compat tests.  But don't call it anywhere here. And continue to remove from interfaces.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 8]
```diff
@@ -231,6 +231,8 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
     protected registry: IFluidDataStoreRegistry | undefined;
 
     protected detachedRuntimeCreation = false;
+    // @ts-expect-error - This shouldn't be referenced in the current version, but needs to be here for back-compat
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:234]: can we mark it with "// back-compat" comment and mention when it can be removed?

**Review by vladsud — APPROVED** [review]

**Comment by markfields** [comment]
The test failure is because I switched a block from NoCompat to FullCompat, which results in an error log with `"eventName": "fluid:telemetry:Container:containerAttach"`.  This is because runWithRetry used to log an error but now logs a generic event.

@anthony-murphy is there support in `itExpects` for different expected logs for different compat versions?  Looks like you can only customize by driver type.  Any suggestions?

File: packages/test/test-end-to-end-tests/src/test/detachedContainerTests.spec.ts [hunk 9]
```diff
@@ -590,7 +590,7 @@ describeFullCompat("Detached Container", (getTestObjectProvider) => {
     });
 });
 
-describeNoCompat("Detached Container", (getTestObjectProvider) => {
+describeFullCompat("Detached Container", (getTestObjectProvider) => {
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/detachedContainerTests.spec.ts:593]:
```suggestion
describeNoCompat("Detached Container", (getTestObjectProvider) => {
```

File: packages/test/test-end-to-end-tests/src/test/detachedContainerTests.spec.ts [hunk 10]
```diff
@@ -31,7 +31,7 @@ import { MessageType, ISummaryTree } from "@fluidframework/protocol-definitions"
 import { DataStoreMessageType } from "@fluidframework/datastore";
 import { ContainerMessageType } from "@fluidframework/container-runtime";
 import { requestFluidObject } from "@fluidframework/runtime-utils";
-import { describeFullCompat, describeNoCompat, itExpects } from "@fluidframework/test-version-utils";
+import { describeFullCompat, itExpects } from "@fluidframework/test-version-utils";
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/detachedContainerTests.spec.ts:34]:
```suggestion
import { describeFullCompat, describeNoCompat, itExpects } from "@fluidframework/test-version-utils";
```

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 11]
```diff
@@ -232,7 +231,8 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
     protected registry: IFluidDataStoreRegistry | undefined;
 
     protected detachedRuntimeCreation = false;
-    public readonly bindToContext: () => void;
+    // @ts-expect-error - This shouldn't be referenced in the current version, but needs to be here for back-compat
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:234]:
```suggestion
    // back-compat (for tests) - can be removed in 2.0.0-alpha.2.0.0, or earlier if compat tests drop n/n-2 coverage
    // @ts-expect-error - This shouldn't be referenced in the current version, but needs to be here for back-compat
```
